pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/array/ListAggFunction.java

```diff
@@ -0,0 +1,155 @@
+    ObjectArrayList<String> valueSet = getObjectArrayList(aggregationResultHolder);
+    BlockValSet blockValSet = blockValSetMap.get(_expression);
+    String[] values = blockValSet.getStringValuesSV();
+    forEachNotNull(length, blockValSet, (from, to) -> {
+      valueSet.addAll(Arrays.asList(values).subList(from, to));
+    });
+  }
+
+  protected ObjectArrayList<String> getObjectArrayList(AggregationResultHolder aggregationResultHolder) {
+    ObjectArrayList<String> valueSet = aggregationResultHolder.getResult();
+    if (valueSet == null) {
+      valueSet = new ObjectArrayList<>();
+      aggregationResultHolder.setValue(valueSet);
+    }
+    return valueSet;
+  }
+
+  protected ObjectArrayList<String> getObjectArrayList(GroupByResultHolder groupByResultHolder, int groupKey) {
+    ObjectArrayList<String> valueSet = groupByResultHolder.getResult(groupKey);
+    if (valueSet == null) {
+      valueSet = new ObjectArrayList<>();
+      groupByResultHolder.setValueForKey(groupKey, valueSet);
+    }
+    return valueSet;
+  }
+
+  @Override
+  public void aggregateGroupBySV(int length, int[] groupKeyArray, GroupByResultHolder groupByResultHolder,
+      Map<ExpressionContext, BlockValSet> blockValSetMap) {
+    BlockValSet blockValSet = blockValSetMap.get(_expression);
+    String[] values = blockValSet.getStringValuesSV();
+    forEachNotNull(length, blockValSet, (from, to) -> {
+      for (int i = from; i < to; i++) {
+        ObjectArrayList<String> groupValueList = getObjectArrayList(groupByResultHolder, groupKeyArray[i]);
+        groupValueList.add(values[i]);
+      }
+    });
+  }
+
+  @Override
+  public void aggregateGroupByMV(int length, int[][] groupKeysArray, GroupByResultHolder groupByResultHolder,
+      Map<ExpressionContext, BlockValSet> blockValSetMap) {
+    BlockValSet blockValSet = blockValSetMap.get(_expression);
+    String[] values = blockValSet.getStringValuesSV();
+    forEachNotNull(length, blockValSet, (from, to) -> {
+      for (int i = from; i < to; i++) {
+        for (int groupKey : groupKeysArray[i]) {
+          ObjectArrayList<String> groupValueList = getObjectArrayList(groupByResultHolder, groupKey);
+          groupValueList.add(values[i]);
+        }
+      }
+    });
+  }
+
+  @Override
+  public ObjectArrayList<String> extractAggregationResult(AggregationResultHolder aggregationResultHolder) {
+    return aggregationResultHolder.getResult();
+  }
+
+  @Override
+  public ObjectArrayList<String> extractGroupByResult(GroupByResultHolder groupByResultHolder, int groupKey) {
+    return groupByResultHolder.getResult(groupKey);
+  }
+
+  @Override
+  public ObjectArrayList<String> merge(ObjectArrayList<String> intermediateResult1,
+      ObjectArrayList<String> intermediateResult2) {
+    if (intermediateResult1 == null) {
+      return intermediateResult2;
+    }
+    if (intermediateResult2 == null) {
+      return intermediateResult1;
+    }
+    intermediateResult1.addAll(intermediateResult2);
```

Comment:
Will the intermediate result ever transferred to another node? If so, this will break because distinct is not honored